library/index.c

```diff
@@ -530,10 +530,14 @@ int eblob_generate_sorted_index(struct eblob_backend *b, struct eblob_base_ctl *
 	struct eblob_map_fd src, dst;
 	int fd, err, len;
 	char *file, *dst_file;
+	int cache_is_empty = 0;
 
 	if (b == NULL || bctl == NULL)
 		return -EINVAL;
 
+	/* check that eblob cache is empty. If cache is empty we can skip iterating and flushing it */
+	cache_is_empty = eblob_cache_empty(b);
+
```

Comment:
Cache says data lives at offset X while it actually lives at offset Y - one user accesses data via cached values another one directly through on-disk lookup.

I'm really concerned about this issue, because I do see a data corruption since 0.22.13 version.
There are 2 commits which are related: this one and indexsort. The latter is indeed buggy - there is a bugreport in Trello, but I turned off that bit for a while. Also, it looks like corruption also happens on servers which did not have indexsort turned on - I can not confirm that though.